pkg/controller/controllerutil/controllerutil.go

```diff
@@ -46,3 +50,75 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
 	object.SetOwnerReferences(append(object.GetOwnerReferences(), ref))
 	return nil
 }
+
+// OperationType is the action result of a CreateOrUpdate call
+type OperationType string
+
+const ( // They should complete the sentence "Deployment default/foo has been ..."
+	// OperationNoop means that the resource has not been changed
+	OperationNoop = "unchanged"
+	// OperationCreated means that a new resource has been created
+	OperationCreated = "created"
+	// OperationUpdated means that an existing resource has been updated
+	OperationUpdated = "updated"
+)
+
+// CreateOrUpdate creates or updates a kubernetes resource. It takes in a key and
+// a placeholder for the existing object and returns the operation executed
+func CreateOrUpdate(ctx context.Context, c client.Client, key client.ObjectKey, existing runtime.Object, t TransformFn) (OperationType, error) {
+	err := c.Get(ctx, key, existing)
+	var obj runtime.Object
+
+	if errors.IsNotFound(err) {
+		// Create a new zero value object so that the in parameter of
+		// TransformFn is always a "clean" object, with only Name and Namespace
+		// set
+		zero := reflect.New(reflect.TypeOf(existing).Elem()).Interface()
```

Comment:
this is what `existing` is for.  It should never be nil, so you shouldn't have to construct a new zero object.

Reply:
The reason for creating a new zero value struct is that you can pass in existing an initialized one, eg. with `.spec` filled. I'm not sure that `client.Get` zeroes the object if it doesn't find one.

What I want to enforce here, is that if this function is going to create a new object, your transform callback always gets an empty object with only Name and Namespace filled in.